Add Hinge metric

ethanwharris opened the pull request:
# Before submitting

- [x] Was this discussed/approved via a Github issue? (no need for typos and docs improvements)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), Pull Request section?
- [x] Did you make sure to update the docs?   
- [x] Did you write any new necessary tests?  

## What does this PR do?
Fixes #89 - added the `multiclass_mode` parameter to enable computation in a one-vs-all fashion (e.g. for multi-class SVMs)

## PR review    
Anyone in the community is free to review the PR once the tests have passed.     
If we didn't discuss your PR in Github issues there's a high chance it will not be merged.

## Did you have fun?
Make sure you had fun coding 🙃

Borda reviewed (commented):
I am not very sure about adding anything named as Loss, could we have it consistent with others?

ethanwharris commented:
@Borda Happy to change if needed :smiley: although not sure what else to call it. I guess the options are:
- `hinge_loss` (consistent with sklearn)
- `hinge` (avoids "loss")
- it's sometimes also called the max-margin objective, so maybe that?

Borda commented:
I would say `hinge` is fine...
cc: @PyTorchLightning/core-metrics other suggestions?

SkafteNicki commented:
+1 for `hinge` :]

Borda commented on docs/source/references/functional.rst:
```diff
@@ -77,6 +77,12 @@ hamming_distance [func]
     :noindex:
 
+hinge [func]
+~~~~~~~~~~~~~~~~~
```
```suggestion
~~~~~~~~~~~~
```

Borda commented on docs/source/references/modules.rst:
```diff
@@ -162,6 +162,18 @@ FBeta
     :noindex:
 
+HammingDistance
+~~~~~~~~~~~~~~~~
```
```suggestion
~~~~~~~~~~~~~~~
```

ethanwharris commented:
@Borda @SkafteNicki Thanks for the comments :smiley: I've added check for `multiclass_mode` and switched hinge loss to just hinge so should all be ready for review now.

Borda commented on torchmetrics/classification/hinge.py:
```diff
@@ -0,0 +1,111 @@
+            dist_sync_fn=dist_sync_fn,
+        )
+
+        self.add_state("loss", default=tensor(0.0), dist_reduce_fx="sum")
```
```suggestion
        self.add_state("measure", default=tensor(0.0), dist_reduce_fx="sum")
```

Borda commented on torchmetrics/classification/hinge.py:
```diff
@@ -0,0 +1,111 @@
+        self.add_state("loss", default=tensor(0.0), dist_reduce_fx="sum")
+        self.add_state("total", default=tensor(0), dist_reduce_fx="sum")
+
+        if multiclass_mode not in (None, 'crammer_singer', 'one_vs_all'):
```
lets define these as Class static StrEnum

Borda commented on torchmetrics/functional/classification/hinge.py:
```diff
@@ -0,0 +1,159 @@
+        mode = DataType.BINARY
+    elif preds.ndim == 2:
+        if preds.shape[0] != target.shape[0]:
+            raise ValueError(
```
seems we do not have tests for raising invalid values, mind add some

ethanwharris replied:
Should be done already here (but codecov is still catching up haha): https://github.com/PyTorchLightning/metrics/blob/cfea3a077801bfb5e4dfa24d2c85a5a7d979d76c/tests/classification/test_hinge.py#L152

Borda reviewed (commented):
very nice, almost there, just last comments from me :]

pep8speaks commented:
Hello @ethanwharris! Thanks for updating this PR. 







* In the file [`torchmetrics/classification/hinge.py`](https://github.com/PyTorchLightning/metrics/blob/d38b8f20dec975cdcb151dfd35eddb68ccfb8d58/torchmetrics/classification/hinge.py):

> [Line 77:1](https://github.com/PyTorchLightning/metrics/blob/d38b8f20dec975cdcb151dfd35eddb68ccfb8d58/torchmetrics/classification/hinge.py#L77): [W293](https://duckduckgo.com/?q=pep8%20W293) blank line contains whitespace

SkafteNicki commented on torchmetrics/classification/hinge.py:
```diff
@@ -0,0 +1,121 @@
+            ``MulticlassMode.ONE_VS_ALL`` or ``"one-vs-all"``.
+
+    Example:
+        >>> import torch
```
```suggestion
        # binary example
        >>> import torch
```

SkafteNicki commented on torchmetrics/classification/hinge.py:
```diff
@@ -0,0 +1,121 @@
+        >>> hinge(preds, target)
+        tensor(0.3000)
+
+        >>> target = torch.tensor([0, 1, 2])
```
```suggestion

        # multiclass example, default mode
        >>> target = torch.tensor([0, 1, 2])
```

SkafteNicki commented on torchmetrics/classification/hinge.py:
```diff
@@ -0,0 +1,121 @@
+        >>> hinge(preds, target)
+        tensor(2.9000)
+
+        >>> target = torch.tensor([0, 1, 2])
```
```suggestion

        # multiclass example, one vs all mode
        >>> target = torch.tensor([0, 1, 2])
```

SkafteNicki commented on torchmetrics/functional/classification/hinge.py:
```diff
@@ -0,0 +1,180 @@
+            ``MulticlassMode.ONE_VS_ALL`` or ``"one-vs-all"``.
+
+    Example:
+        >>> import torch
```
above comments also applies to these examples

ethanwharris replied:
Done :)

SkafteNicki reviewed (approved):
smaller comments else LGTM

ethanwharris commented:
@PyTorchLightning/core-metrics I was having a weird issue here where the test was hanging forever rather than reporting an error with `ddp` and `dist_sync_on_step` both `True`. Managed to fix the bug so all works now, but ideally it would report the test failure rather than hang. Is this a known issue?
